src/NHibernate/Loader/Criteria/CriteriaQueryTranslator.cs

```diff
@@ -26,6 +26,7 @@ public class CriteriaQueryTranslator : ICriteriaQuery
 		private readonly string rootEntityName;
 		private readonly string rootSQLAlias;
 		private int indexForAlias = 0;
+		private IList<EntityProjection> _entityProjections = new List<EntityProjection>();
```

Comment:
Also, please make field type of `List` instead of `IList` to avoid unnecessary boxing.

Reply:
@hazzik Ok. But could you please clarify what boxing do you expect to see? I've just checked with Resharper IL viewer and don't see any difference in initializing both `IList<EntityProjection>` or `List<EntityProjection>`.  `IList<EntityProjection>` is directly implemented by `List<EntityProjection>` - no (co/contr)-variant conversion needed.

And from what I see I can directly assign `IEnumerable<object> objs = new List<EntityProjection>();` without any penalties with implicit hidden conversions.

Reply:
The enumerator of `List` is declared as `struct` and exposed as bare type. Both `IEnumerable<>.GeEnumerator` and `IEnumerable.GetEnumerator` implemented as explicit and there will be boxing of enumerator when it is accessed from `IEnumerable.GetEnumerator`

Reply:
https://marcinjuraszek.com/2013/10/playing-around-with-listt-part-two-ienumerable-and-ienumerablet-implementation.html

Reply:
@hazzik Interesting! Thanks. So it's about the price of single boxing per each enumeration via `foreach`. But that means that I also need to expose direct `List<>` class in `GetEntityProjections()` method and `InitProjection` method should also expect direct `List<>` class:

`protected void InitProjection(..., *List<EntityProjection> entityProjections*)`

Did I get it right? Are you sure you want me to go this way (except field type changing  - it's done already)? Considering that there is no enumeration via foreach/LINQ in  current implementation.

Also it seems it helps to avoid boxing only if `foreach` is used on list. It's still boxed for any LINQ calls (which is logical but I was hoping to see some CLR magic :) ):
https://dotnetfiddle.net/voVECl

Reply:
No, just field.